Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,230 +1,244 @@
+        //find next free space  
+        while(hashTable->items[index] != NULL && hashTable->items[index]->key != key && hashTable->items[index]->key != -1) 
+        { 
+            index++; 
+            index % hashTable->size; 
+        } 
+    }
+    hashTable->items[index] = createItem(key, value, &success);
+    if (hashTable->items[index]->value != value) {
+        return ERROR;
+    }
+    return SUCCESS;
 }
-//To-do: will need to think of how to handle collisions 
-// buildHashTable receives an array of keys and an array of values
-// key[i] will be mapped to value[i] when building the hashTable
-// loops through the values to create a key for the current value then inserts it into the existing hashTable.
+
+// bulkInsert 
 // returns a SUCCESS code if all values are added into the hashTable successfully 
-int buildHashTable(int type, int *keys, int *values) {
-    int errcode = SUCCESS;
-    for (int i = 0; i < SIZE; i++) {
-        int key = hashing(keys[i]);
-        insert(type, key, values[i], errcode);
-   }
-   return errcode;
+// returns an ERROR code if a value cannot be added into the hashTable 
+int bulkInsert(HashTable * hashTable, char keys[][10], int* values) {
+    int success = NULL;
+    for (int i = 0; i < hashTable->size; i++) {
+        success = insert(hashTable , keys[i], values[i]);
+        if (success == ERROR) {
+            return ERROR;
+        }
+    }
+   return SUCCESS;
 }
 
-//to-do: Will need to think of a rescue in the case of a collision insertion
-// searchHashTable takes in an integer key
-// calls the helper function to hash the received key to correspond to an expected index within the hash table
-// if the key at the position in the test hashTable indicated by the hashing function matches the received key then return a SUCCESS code
-// if a mistmatch occurs return an ERROR code
-int searchHashTable(int type, int key, int * success) {
-    if (key > SIZE || key < 0) return ERROR;
-    int hashIndex = hashing(key);
-    if (testHash[hashIndex]->key == key) {
-        *success = SUCCESS;
+// createTestHashTable
+// returns a HashTable struct to use for testing 
+HashTable * createTestHashTable() {
+    char keys[10][10] = { "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen"};
+    int value[10]= { 10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
+    HashTable * testHash = createHashTable(10);
+    bulkInsert(testHash, keys, value);
+    return testHash;
+}
+
+// searchHashTable
+// if the key/value pair is found through linear probing then return a SUCCESS code
+// if key/value pair is not found return an ERROR code
+int searchHashTable(HashTable * hashTable, char key[], int value) {
+    int hashIndex = hashing(hashTable,key);
+    for (int i = 0; i < hashTable->size; i++) {
+        int index = (hashIndex + i) % hashTable->size;
+        if (hashTable->items[index]->value == value && strcmp(hashTable->items[index]->key, key) == 0) {
```

Comment:
Do we need to check that `hashTable->items[index] !=  NULL` before dereferencing `->key` and `->value`?

Reply:
Can you clarify your question? I am not following why to do that here.

Reply:
Updated per meeting feedback to check index before key and value! [commit 9ae550b437f22e489e3fe6fa42559c5e9c6e3f90](https://github.com/kcarrel/Training-Notes/pull/12/commits/9ae550b437f22e489e3fe6fa42559c5e9c6e3f90)